File: app/src/main/java/com/bopr/android/smailer/ui/EditFilterListItemDialogFragment.java
package com.bopr.android.smailer.ui;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentActivity;

/**
 * Base filter list item editor dialog.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public abstract class EditFilterListItemDialogFragment extends DialogFragment {

    private OnClose onClose;
    private int title;

    public void setTitle(int title) {
        this.title = title;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onDestroyView() {
        /* avoiding of disappearing on rotation */
        if (getDialog() != null && getRetainInstance()) {
            getDialog().setDismissMessage(null);
        }
        super.onDestroyView();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = getDialog();
        if (dialog == null) {
            View view = createView();
            dialog = new AlertDialog.Builder(requireContext())
                    .setTitle(title)
                    .setView(view)
                    .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {

                        @Override
                        public void onClick(DialogInterface dialogInterface, int which) {
                            onClose.onOkClick(getValue());
                        }
                    })
                    .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {

                        @Override
                        public void onClick(DialogInterface dialogInterface, int which) {
                            dialogInterface.cancel();
                        }
                    })
                    .create();

            /* show soft keyboard when dialog is open */
            //noinspection ConstantConditions
            dialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        }
        return dialog;
    }

    protected abstract String getValue();

    @NonNull
    protected abstract String createTag();

    @NonNull
    protected abstract View createView();

    void showDialog(FragmentActivity activity) {
        show(activity.getSupportFragmentManager(), createTag());
    }

    void setOnClose(OnClose onClose) {
        this.onClose = onClose;
    }

    public interface OnClose {

        void onOkClick(String result);

    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/EditFilterListItemDialogFragment.kt
package com.bopr.android.smailer.ui

import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity

/**
 * Base filter list item editor dialog.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
abstract class EditFilterListItemDialogFragment : DialogFragment() {

    private var onOkClicked: ((String?) -> Unit)? = null
    private var titleRes = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        var dialog = dialog
        if (dialog == null) {
            val view = createView()
            dialog = AlertDialog.Builder(requireContext())
                    .setTitle(titleRes)
                    .setView(view)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        onOkClicked?.invoke(getValue())
                    }
                    .setNegativeButton(android.R.string.cancel) { dialogInterface, _ ->
                        dialogInterface.cancel()
                    }
                    .create()

            /* show soft keyboard when dialog is open */
            dialog.getWindow()!!.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
        }
        return dialog
    }

    override fun onDestroyView() {
        /* avoiding of disappearing on rotation */
        if (dialog != null && retainInstance) {
            dialog!!.setDismissMessage(null)
        }
        super.onDestroyView()
    }

    protected abstract fun getValue(): String?

    protected abstract fun createTag(): String

    protected abstract fun createView(): View

    fun setTitle(@StringRes titleRes: Int) {
        this.titleRes = titleRes
    }

    fun setOnOkClicked(action: (String?) -> Unit) {
        onOkClicked = action
    }

    fun showDialog(activity: FragmentActivity) {
        show(activity.supportFragmentManager, createTag())
    }
}